src/problem.cpp

```diff
@@ -0,0 +1,333 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/problem.hpp>
+
+#include <miopen/conv/problem_description.hpp>
+#include <miopen/convolution.hpp>
+#include <miopen/conv_algo_name.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/solution.hpp>
+#include <miopen/search_options.hpp>
+
+#include <nlohmann/json.hpp>
+
+namespace miopen {
+
+namespace detail {
+
+// Selected only with empty VariantArgs
+template <int i, template <class Type> class Visitor, class... VariantArgs>
+struct VisitTypeImpl
+{
+    template <class... Args>
+    void operator()(int, Args...)
+    {
+        MIOPEN_THROW(miopenStatusInvalidValue);
+    }
+};
+
+template <int i, template <class Type> class Visitor, class VariantArg, class... VariantArgs>
+struct VisitTypeImpl<i, Visitor, VariantArg, VariantArgs...>
+{
+    template <class... Args>
+    void operator()(int id, Args... args)
+    {
+        if(i == id)
+        {
+            Visitor<VariantArg>{args...}();
+            return;
+        }
+
+        VisitTypeImpl<i + 1, Visitor, VariantArgs...>{}(id, args...);
+    }
+};
+
+template <template <class Type> class Visitor, class... VariantArgs>
+struct VisitType;
+
+template <template <class Type> class Visitor, class... VariantArgs>
+struct VisitType<Visitor, boost::variant<VariantArgs...>>
+{
+    template <class... Args>
+    void operator()(int id, Args... args)
+    {
+        detail::VisitTypeImpl<0, Visitor, VariantArgs...>{}(id, args...);
+    }
+};
+
+} // namespace detail
+
+template <template <class Type> class Visitor, class Variant, class... Args>
+void VisitType(int id, Args... args)
+{
+    detail::VisitType<Visitor, Variant>{}(id, args...);
+}
+
+namespace detail {
+
+struct ProblemFindVariantVisitor
+{
+    const Problem* problem;
+    Handle* handle;
+    const SearchOptions* options;
+    std::size_t max_solutions;
+
+    std::vector<Solution> operator()(const ConvolutionDescriptor& conv) const
+    {
+        return problem->FindConvSolutions(*handle, *options, max_solutions, conv);
+    }
+
+    template <class Unimplemented>
+    std::vector<Solution> operator()(const Unimplemented&) const
+    {
+        MIOPEN_THROW(miopenStatusNotImplemented);
+        return {};
+    }
+};
+
+} // namespace detail
+
+std::vector<Solution> Problem::FindSolutions(Handle& handle,
+                                             const SearchOptions& options,
+                                             std::size_t max_solutions) const
+{
+    const auto visitor = detail::ProblemFindVariantVisitor{this, &handle, &options, max_solutions};
```

Comment:
I think using an overloaded lambdas would be simpler. With boost, you can use [`boost::hof::match`](https://www.boost.org/doc/libs/1_72_0/libs/hof/doc/html/include/boost/hof/match.html) or [`boost::hof::first_of`](https://www.boost.org/doc/libs/1_72_0/libs/hof/doc/html/include/boost/hof/first_of.html). In C++17 without boost, you can add this:

```cpp
template <class... Fs>
struct overloaded : Fs...
{
    using Fs::operator()...;
    overloaded(Fs... fs) : Fs(fs)... {}
};

template <class... Fs>
overloaded<Fs...> overload(Fs... fs)
{
    return {fs...};
}
```